crates/ruff_python_formatter/src/statement/stmt_with.rs

```diff
@@ -1,12 +1,54 @@
-use crate::{not_yet_implemented, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::comments::trailing_comments;
+use crate::prelude::*;
+use crate::{FormatNodeRule, PyFormatter};
+use ruff_formatter::{format_args, write, Buffer, FormatResult};
+use ruff_python_ast::prelude::*;
 use rustpython_parser::ast::StmtWith;
 
 #[derive(Default)]
 pub struct FormatStmtWith;
 
 impl FormatNodeRule<StmtWith> for FormatStmtWith {
     fn fmt_fields(&self, item: &StmtWith, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [not_yet_implemented(item)])
+        let StmtWith {
+            range: _,
+            items,
+            body,
+            type_comment: _,
+        } = item;
+
+        let comments = f.context().comments().clone();
+        let dangling_comments = comments.dangling_comments(item.as_any_node_ref());
+        let trailing_items_comments = body.first().map(|body_start| {
+            let before_body = dangling_comments
+                .partition_point(|comment| comment.slice().end() < body_start.start());
+
+            let (trailing_items_comments, rest) = dangling_comments.split_at(before_body);
+            debug_assert!(rest.is_empty());
+            trailing_items_comments
+        });
+
+        let joined_items = format_with(|f| f.join_comma_separated().nodes(items.iter()).finish());
+
+        write!(
+            f,
+            [
+                text("with"),
+                space(),
+                group(&format_args![
+                    if_group_breaks(&text("(")),
+                    soft_block_indent(&joined_items),
+                    if_group_breaks(&text(")")),
+                ]),
```

Comment:
Nit: You can use `optional_parentheses(&joined_items)` if this lands after #5328